server/app/auth/ApiAuthenticator.java

```diff
@@ -0,0 +1,97 @@
+package auth;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import java.time.Instant;
+import javax.inject.Inject;
+import javax.inject.Provider;
+import models.ApiKey;
+import org.apache.commons.net.util.SubnetUtils;
+import org.pac4j.core.context.WebContext;
+import org.pac4j.core.context.session.SessionStore;
+import org.pac4j.core.credentials.Credentials;
+import org.pac4j.core.credentials.UsernamePasswordCredentials;
+import org.pac4j.core.credentials.authenticator.Authenticator;
+import org.pac4j.core.exception.BadCredentialsException;
+import play.cache.NamedCache;
+import play.cache.SyncCacheApi;
+import services.apikey.ApiKeyService;
+
+/**
+ * Authenticator for API requests based on HTTP basic auth and backed by the {@link ApiKey}
+ * resource. Background: https://en.wikipedia.org/wiki/Basic_access_authentication
+ *
+ * <p>When referenced by a request, {@code ApiKey}s are stored in the "api-keys" named cache, keyed
+ * by their key ID. This allows controller code to load the key without the need for a subsequent
+ * database query. This is necessary because pac4j creates profiles in the next step, but we need
```

Comment:
"This is necessary because..."

I'm not following how this is a justification for the cache, especially since we read from the DB and write to the cache from this code which is seemingly ok despite the comment that profiles coming next is a problem.

Reply:
update the last sentence of this paragraph to "This is necessary because pac4j creates profiles (which are what's made available to controller code through the request object) in the next step, but we need the API key to perform authentication (this step). The purpose of the cache is to avoid making multiple database calls to retrieve the API key throughout the cycle of the request."